### ISSUE TITLE
[BUG] Integration Tests Keep Failing

### ISSUE THREAD

**Issue by steven-murray** [21cmfast/21cmFAST#133]
**Describe the bug:**

Integration tests seem to fail reasonably randomly, seemingly due to rounding error. This is a problem because it's super annoying to debug, and wastes time.

**Expected behavior:**

They should pass consistently from machine to machine.

This issue is a place to discuss how we might go about making the tests more robust to this kind of error.

**Comment by BradGreig** [comment]
The obvious thing for me, and what I do a lot when adding new features etc., is to turn off the RNG for the inital conditions. This way, serial/parallel code calls remain the same, and outputs can be easily compared across versions etc. 

I noted numerical differences on 21-cm PS and globally averaged quantities across versions when generating summary statistics. However, when turning them off, the results were exact. To me, I am not 100% confident that this means a rounding error, or if there is some other small error propagating through that is difficult to catch. I've spent a considerable time exploring this, but haven't come up with much.

**Comment by andreimesinger** [comment]
one can put in an error tolerance threshold

**Comment by steven-murray** [comment]
Just wait -- why is there a difference between having no RNG and having an RNG with a specified seed?

@andreimesinger we do have an error threshold: Absolute tolerance of 1e-5 and Relative tol of 1e-2 (for coeval) and 1e-3 (for lightcone). 

Sometimes when adding new integration tests (these tests are basically comparing power spectra of output coeval/lightcone boxes by computing the box locally then making sure the cloud CI machine computes the same thing) they will not match, seemingly just because a different machine ran it. Errors can be up to a few percent at the higher k's. We could of course just set the tolerance a bit higher every time something breaks it, but that does not sound like a good idea (and a bit like cheating...). 

@BradGreig how does one turn off the RNG completely?

BTW I get exact correspondence when running the test on the same machine that produced the data in the first place (even with RNG). So it's not "random" error in the sense that the same machine will give random results every run. Just that different machines seem to give slightly different answers.

**Comment by BradGreig** [comment]
@steven-murray, yeah, I confirm that performing on the same machine I always get exact results. Issues always arise when pushing to travis. 

What I noticed was that as I implemented the `hires-rho` branch, this changed the order of some of the calculation (i.e. FFTs in different places etc.). Which meant cumulative rounding errors were different (as there seems to be rounding errors that arise from the FFTs). Yet, when turning off the RNGs this completely went away.

By turning off the RNGs, I mean I manually comment out the RNGs for the complex numbers in the initial conditions. Perhaps instead this should become a feature of the code, that when run in debug mode RNGs are automatically turned off. Or, perhaps its better to have a user changeable flag (maybe in global_params as it isn't to be used often) in order for it to be easily implemented in tests.
